{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get sequences\n",
    "of the species that are in the numt dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from bold import boldAPI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "dna_data_path = os.path.join('data', 'dna_sequences.csv')\n",
    "if not os.path.exists(dna_data_path):\n",
    "    dna_df = pd.read_csv(dna_data_path)\n",
    "    unique_species = dna_df['Species'].unique()\n",
    "    vfunc = np.vectorize(boldAPI.get_sequences)\n",
    "    results = pd.concat(vfunc(unique_species))\n",
    "else: \n",
    "    results = pd.read_csv(dna_data_path)\n",
    "\n",
    "results = results[results['Sequence'].str.contains('N') == False] # if there is an N its an error\n",
    "results['Sequence'] = results['Sequence'].apply(lambda x: x.replace('-', '')) # we dont care about the - its an alignment artifact and we can just remove them\n",
    "results = results.drop_duplicates(subset='Sequence') # there are a lot of duplicate sequences just drop all but the first\n",
    "\n",
    "results.to_csv(dna_data_path, index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Acropora millepora\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>BOLDProcessId</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBCI10386-19</td>\n",
       "      <td>TACGTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE20474-24</td>\n",
       "      <td>TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE25693-24</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE26948-24</td>\n",
       "      <td>TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE28464-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE31107-24</td>\n",
       "      <td>TGCTTCTACCCAATTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE35612-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE43179-24</td>\n",
       "      <td>TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE45768-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACACGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE47451-24</td>\n",
       "      <td>GTTTTTACCCGGTTATTCGGGTGAGTTGGAGTGGTCGCGGCCTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE4891-24</td>\n",
       "      <td>TGCTTCTACCCATGGAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE50146-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE53137-24</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE67085-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE73842-24</td>\n",
       "      <td>TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE74315-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE7881-24</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE87666-24</td>\n",
       "      <td>TGCTCTACCCATTCAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF17245-24</td>\n",
       "      <td>TGCTTTTACCCGGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF54948-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF54992-24</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF71523-24</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV35301-23</td>\n",
       "      <td>TTGGAGTTCGGGCCGCACGCGCGCTACACTGAGCGCCGCAGCGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV40195-24</td>\n",
       "      <td>TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV40196-24</td>\n",
       "      <td>TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5136-21</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5137-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5138-21</td>\n",
       "      <td>TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5139-21</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5140-21</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5141-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5142-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5143-21</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5144-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5145-21</td>\n",
       "      <td>TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5146-21</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5147-21</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5148-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5149-21</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5150-21</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5151-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5152-21</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5153-21</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5154-21</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>OGLA1208-11</td>\n",
       "      <td>---GTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>OGLA1851-11</td>\n",
       "      <td>---------------------------GGCATGATTGGCACGGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>SCL034-22</td>\n",
       "      <td>CATAAAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Species  BOLDProcessId  \\\n",
       "0  Acropora millepora   GBCI10386-19   \n",
       "0  Acropora millepora  GBFSE20474-24   \n",
       "0  Acropora millepora  GBFSE25693-24   \n",
       "0  Acropora millepora  GBFSE26948-24   \n",
       "0  Acropora millepora  GBFSE28464-24   \n",
       "0  Acropora millepora  GBFSE31107-24   \n",
       "0  Acropora millepora  GBFSE35612-24   \n",
       "0  Acropora millepora  GBFSE43179-24   \n",
       "0  Acropora millepora  GBFSE45768-24   \n",
       "0  Acropora millepora  GBFSE47451-24   \n",
       "0  Acropora millepora   GBFSE4891-24   \n",
       "0  Acropora millepora  GBFSE50146-24   \n",
       "0  Acropora millepora  GBFSE53137-24   \n",
       "0  Acropora millepora  GBFSE67085-24   \n",
       "0  Acropora millepora  GBFSE73842-24   \n",
       "0  Acropora millepora  GBFSE74315-24   \n",
       "0  Acropora millepora   GBFSE7881-24   \n",
       "0  Acropora millepora  GBFSE87666-24   \n",
       "0  Acropora millepora  GBFSF17245-24   \n",
       "0  Acropora millepora  GBFSF54948-24   \n",
       "0  Acropora millepora  GBFSF54992-24   \n",
       "0  Acropora millepora  GBFSF71523-24   \n",
       "0  Acropora millepora  GBINV35301-23   \n",
       "0  Acropora millepora  GBINV40195-24   \n",
       "0  Acropora millepora  GBINV40196-24   \n",
       "0  Acropora millepora   GBMNE5136-21   \n",
       "0  Acropora millepora   GBMNE5137-21   \n",
       "0  Acropora millepora   GBMNE5138-21   \n",
       "0  Acropora millepora   GBMNE5139-21   \n",
       "0  Acropora millepora   GBMNE5140-21   \n",
       "0  Acropora millepora   GBMNE5141-21   \n",
       "0  Acropora millepora   GBMNE5142-21   \n",
       "0  Acropora millepora   GBMNE5143-21   \n",
       "0  Acropora millepora   GBMNE5144-21   \n",
       "0  Acropora millepora   GBMNE5145-21   \n",
       "0  Acropora millepora   GBMNE5146-21   \n",
       "0  Acropora millepora   GBMNE5147-21   \n",
       "0  Acropora millepora   GBMNE5148-21   \n",
       "0  Acropora millepora   GBMNE5149-21   \n",
       "0  Acropora millepora   GBMNE5150-21   \n",
       "0  Acropora millepora   GBMNE5151-21   \n",
       "0  Acropora millepora   GBMNE5152-21   \n",
       "0  Acropora millepora   GBMNE5153-21   \n",
       "0  Acropora millepora   GBMNE5154-21   \n",
       "0  Acropora millepora    OGLA1208-11   \n",
       "0  Acropora millepora    OGLA1851-11   \n",
       "0  Acropora millepora      SCL034-22   \n",
       "\n",
       "                                            Sequence  \n",
       "0  TACGTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...  \n",
       "0  TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAATTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACACGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GTTTTTACCCGGTTATTCGGGTGAGTTGGAGTGGTCGCGGCCTGCC...  \n",
       "0  TGCTTCTACCCATGGAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTCTACCCATTCAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGCG...  \n",
       "0  TGCTTTTACCCGGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TTGGAGTTCGGGCCGCACGCGCGCTACACTGAGCGCCGCAGCGAGT...  \n",
       "0  TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...  \n",
       "0  TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  ---GTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...  \n",
       "0  ---------------------------GGCATGATTGGCACGGCCT...  \n",
       "0  CATAAAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAG...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from bold import boldAPI\n",
    "import importlib\n",
    "importlib.reload(boldAPI)\n",
    "\n",
    "boldAPI.get_sequences('Acropora millepora')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>BOLDProcessId</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abscondita terminalis</td>\n",
       "      <td>GBCL39854-19</td>\n",
       "      <td>CTTAATAAATGACTTTTTTCAACAAATCATAAAGACATTGGAACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abscondita terminalis</td>\n",
       "      <td>SPMIS002-22</td>\n",
       "      <td>AACTTTATATTTTATCTTCGGAATCTGAGCAGGAATATTAGGAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Actias luna luna</td>\n",
       "      <td>GBGLS176-19</td>\n",
       "      <td>TTTTGATTTTTTGGGCATCCAGAAGTTTATATTTTAATTTTACCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actias luna</td>\n",
       "      <td>LGSM242-04</td>\n",
       "      <td>AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Actias luna</td>\n",
       "      <td>LGSMC795-05</td>\n",
       "      <td>AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41944</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMTG5022-16</td>\n",
       "      <td>ATGTTAATATTTTTTTATTTTTCTACCTTTTTTCTTTGTGGTTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41945</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMTG5022-16</td>\n",
       "      <td>ATGAATAAACCCATACGAAAAAACCACCCACTAATCAAAATTATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41946</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMH10124-19</td>\n",
       "      <td>GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41947</th>\n",
       "      <td>Zootermopsis nevadensis nuttingi</td>\n",
       "      <td>GBMH9919-19</td>\n",
       "      <td>GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41948</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>SPMIS419-22</td>\n",
       "      <td>AACATTATACTTCATCTTCGGGGCATGATCCGGAATATTAGGAACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40565 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Species BOLDProcessId  \\\n",
       "0                 Abscondita terminalis  GBCL39854-19   \n",
       "1                 Abscondita terminalis   SPMIS002-22   \n",
       "2                      Actias luna luna   GBGLS176-19   \n",
       "3                           Actias luna    LGSM242-04   \n",
       "4                           Actias luna   LGSMC795-05   \n",
       "...                                 ...           ...   \n",
       "41944           Zootermopsis nevadensis  GBMTG5022-16   \n",
       "41945           Zootermopsis nevadensis  GBMTG5022-16   \n",
       "41946           Zootermopsis nevadensis  GBMH10124-19   \n",
       "41947  Zootermopsis nevadensis nuttingi   GBMH9919-19   \n",
       "41948           Zootermopsis nevadensis   SPMIS419-22   \n",
       "\n",
       "                                                Sequence  \n",
       "0      CTTAATAAATGACTTTTTTCAACAAATCATAAAGACATTGGAACTT...  \n",
       "1      AACTTTATATTTTATCTTCGGAATCTGAGCAGGAATATTAGGAACC...  \n",
       "2      TTTTGATTTTTTGGGCATCCAGAAGTTTATATTTTAATTTTACCAG...  \n",
       "3      AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...  \n",
       "4      AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...  \n",
       "...                                                  ...  \n",
       "41944  ATGTTAATATTTTTTTATTTTTCTACCTTTTTTCTTTGTGGTTTGT...  \n",
       "41945  ATGAATAAACCCATACGAAAAAACCACCCACTAATCAAAATTATAA...  \n",
       "41946  GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...  \n",
       "41947  GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...  \n",
       "41948  AACATTATACTTCATCTTCGGGGCATGATCCGGAATATTAGGAACA...  \n",
       "\n",
       "[40565 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get amino acid sequences for the DNA seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from plmcluster import proteinextraction\n",
    "# sequencePath = os.path.join('data', 'BOLD_sequences.csv')\n",
    "# if os.path.exists(sequencePath) and:\n",
    "#     results = pd.read_csv(sequencePath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jnicolow/.conda/envs/numt/lib/python3.11/site-packages/Bio/Seq.py:2334: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "results.drop_duplicates(subset='Sequence', inplace=True) # there is a chance that after N and - being removed there are sequences that are duplicates now\n",
    "results['AminoAcidSeq'] = results['Sequence'].apply(proteinextraction.get_aminoacid_seq)\n",
    "results.to_csv(os.path.join('data', 'BOLD_seqs_with_protein.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
