{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from bold import boldAPI"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Sequences\n",
    "if dna_sequences.csv exists then load it otherwise download species in `bold/bold_queries/species_names.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading: Acropora millepora\n",
      "You have exceeded your allowed request quota. If you wish to download large volume of data, please contact support@boldsystems.org for instruction on the process. \n",
      "downloading: Acropora millepora\n",
      "You have exceeded your allowed request quota. If you wish to download large volume of data, please contact support@boldsystems.org for instruction on the process. \n",
      "downloading: pocillopora damicornis\n",
      "You have exceeded your allowed request quota. If you wish to download large volume of data, please contact support@boldsystems.org for instruction on the process. \n",
      "downloading: favia fragum\n",
      "You have exceeded your allowed request quota. If you wish to download large volume of data, please contact support@boldsystems.org for instruction on the process. \n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "Cannot save file into a non-existent directory: 'data'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[5], line 13\u001b[0m\n\u001b[0;32m     10\u001b[0m results[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m results[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mapply(\u001b[38;5;28;01mlambda\u001b[39;00m x: x\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)) \u001b[38;5;66;03m# we dont care about the - its an alignment artifact and we can just remove them\u001b[39;00m\n\u001b[0;32m     11\u001b[0m results \u001b[38;5;241m=\u001b[39m results\u001b[38;5;241m.\u001b[39mdrop_duplicates(subset\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msequence\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;66;03m# there are a lot of duplicate sequences just drop all but the first\u001b[39;00m\n\u001b[1;32m---> 13\u001b[0m \u001b[43mresults\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdna_data_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\jcnic\\anaconda3\\envs\\numt\\Lib\\site-packages\\pandas\\core\\generic.py:3772\u001b[0m, in \u001b[0;36mNDFrame.to_csv\u001b[1;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[0;32m   3761\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[0;32m   3763\u001b[0m formatter \u001b[38;5;241m=\u001b[39m DataFrameFormatter(\n\u001b[0;32m   3764\u001b[0m     frame\u001b[38;5;241m=\u001b[39mdf,\n\u001b[0;32m   3765\u001b[0m     header\u001b[38;5;241m=\u001b[39mheader,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   3769\u001b[0m     decimal\u001b[38;5;241m=\u001b[39mdecimal,\n\u001b[0;32m   3770\u001b[0m )\n\u001b[1;32m-> 3772\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   3773\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3774\u001b[0m \u001b[43m    \u001b[49m\u001b[43mlineterminator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlineterminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3775\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3776\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3777\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3778\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3779\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3780\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3781\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3782\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3783\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3784\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3785\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3786\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3787\u001b[0m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3788\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   3789\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\jcnic\\anaconda3\\envs\\numt\\Lib\\site-packages\\pandas\\io\\formats\\format.py:1186\u001b[0m, in \u001b[0;36mDataFrameRenderer.to_csv\u001b[1;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, lineterminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[0;32m   1165\u001b[0m     created_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m   1167\u001b[0m csv_formatter \u001b[38;5;241m=\u001b[39m CSVFormatter(\n\u001b[0;32m   1168\u001b[0m     path_or_buf\u001b[38;5;241m=\u001b[39mpath_or_buf,\n\u001b[0;32m   1169\u001b[0m     lineterminator\u001b[38;5;241m=\u001b[39mlineterminator,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1184\u001b[0m     formatter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfmt,\n\u001b[0;32m   1185\u001b[0m )\n\u001b[1;32m-> 1186\u001b[0m \u001b[43mcsv_formatter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1188\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[0;32m   1189\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "File \u001b[1;32mc:\\Users\\jcnic\\anaconda3\\envs\\numt\\Lib\\site-packages\\pandas\\io\\formats\\csvs.py:240\u001b[0m, in \u001b[0;36mCSVFormatter.save\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    236\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    237\u001b[0m \u001b[38;5;124;03mCreate the writer & save.\u001b[39;00m\n\u001b[0;32m    238\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    239\u001b[0m \u001b[38;5;66;03m# apply compression and byte/text conversion\u001b[39;00m\n\u001b[1;32m--> 240\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    241\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    242\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    243\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    244\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    245\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    246\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    247\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mas\u001b[39;00m handles:\n\u001b[0;32m    248\u001b[0m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[0;32m    249\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwriter \u001b[38;5;241m=\u001b[39m csvlib\u001b[38;5;241m.\u001b[39mwriter(\n\u001b[0;32m    250\u001b[0m         handles\u001b[38;5;241m.\u001b[39mhandle,\n\u001b[0;32m    251\u001b[0m         lineterminator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlineterminator,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    256\u001b[0m         quotechar\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquotechar,\n\u001b[0;32m    257\u001b[0m     )\n\u001b[0;32m    259\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_save()\n",
      "File \u001b[1;32mc:\\Users\\jcnic\\anaconda3\\envs\\numt\\Lib\\site-packages\\pandas\\io\\common.py:737\u001b[0m, in \u001b[0;36mget_handle\u001b[1;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[0;32m    735\u001b[0m \u001b[38;5;66;03m# Only for write methods\u001b[39;00m\n\u001b[0;32m    736\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode \u001b[38;5;129;01mand\u001b[39;00m is_path:\n\u001b[1;32m--> 737\u001b[0m     \u001b[43mcheck_parent_directory\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    739\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m compression:\n\u001b[0;32m    740\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m compression \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mzstd\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    741\u001b[0m         \u001b[38;5;66;03m# compression libraries do not like an explicit text-mode\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\jcnic\\anaconda3\\envs\\numt\\Lib\\site-packages\\pandas\\io\\common.py:600\u001b[0m, in \u001b[0;36mcheck_parent_directory\u001b[1;34m(path)\u001b[0m\n\u001b[0;32m    598\u001b[0m parent \u001b[38;5;241m=\u001b[39m Path(path)\u001b[38;5;241m.\u001b[39mparent\n\u001b[0;32m    599\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m parent\u001b[38;5;241m.\u001b[39mis_dir():\n\u001b[1;32m--> 600\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(\u001b[38;5;124mrf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot save file into a non-existent directory: \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mparent\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mOSError\u001b[0m: Cannot save file into a non-existent directory: 'data'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "dna_data_path = os.path.join('data', 'dna_sequences.csv')\n",
    "if not os.path.exists(dna_data_path):\n",
    "    species_names = list(set(boldAPI.species_from_species_namestxt())) # set to make sure its unique\n",
    "    vfunc = np.vectorize(boldAPI.get_sequences)\n",
    "    results = pd.concat(vfunc(species_names))\n",
    "else: \n",
    "    results = pd.read_csv(dna_data_path)\n",
    "\n",
    "results = results[results['sequence'].str.contains('N') == False] # if there is an N its an error\n",
    "results['sequence'] = results['sequence'].apply(lambda x: x.replace('-', '')) # we dont care about the - its an alignment artifact and we can just remove them\n",
    "results = results.drop_duplicates(subset='sequence') # there are a lot of duplicate sequences just drop all but the first\n",
    "\n",
    "results.to_csv(dna_data_path, index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Australia\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "process_id = 'GBFSE25693-24'\n",
    "url = f'https://www.boldsystems.org/index.php/API_Public/specimen?ids={process_id}'\n",
    "\n",
    "\n",
    "response = requests.get(url)\n",
    "\n",
    "country_string = '</country'\n",
    "filtered_list = [item for item in response.text.split('>')[1:] if country_string in item]\n",
    "print(filtered_list[0].replace(country_string, ''))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading: Acropora millepora\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>BOLD_process_id</th>\n",
       "      <th>country</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBCI10386-19</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TACGTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE20474-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE25693-24</td>\n",
       "      <td>Australia</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE26948-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE28464-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE31107-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAATTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE35612-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE43179-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE45768-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACACGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE47451-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>GTTTTTACCCGGTTATTCGGGTGAGTTGGAGTGGTCGCGGCCTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE4891-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCATGGAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE50146-24</td>\n",
       "      <td>Papua New Guinea</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE53137-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE67085-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE73842-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE74315-24</td>\n",
       "      <td>Papua New Guinea</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE7881-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSE87666-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTCTACCCATTCAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF17245-24</td>\n",
       "      <td>Papua New Guinea</td>\n",
       "      <td>TGCTTTTACCCGGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF54948-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF54992-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBFSF71523-24</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV35301-23</td>\n",
       "      <td>Unrecoverable</td>\n",
       "      <td>TTGGAGTTCGGGCCGCACGCGCGCTACACTGAGCGCCGCAGCGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV40195-24</td>\n",
       "      <td>Indonesia</td>\n",
       "      <td>TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBINV40196-24</td>\n",
       "      <td>Indonesia</td>\n",
       "      <td>TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5136-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5137-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5138-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5139-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5140-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5141-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5142-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5143-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5144-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5145-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5146-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5147-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5148-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5149-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5150-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5151-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5152-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5153-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>GBMNE5154-21</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>OGLA1208-11</td>\n",
       "      <td>Australia</td>\n",
       "      <td>---GTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>OGLA1851-11</td>\n",
       "      <td>Australia</td>\n",
       "      <td>---------------------------GGCATGATTGGCACGGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acropora millepora</td>\n",
       "      <td>SCL034-22</td>\n",
       "      <td>Malaysia</td>\n",
       "      <td>CATAAAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              species BOLD_process_id           country  \\\n",
       "0  Acropora millepora    GBCI10386-19           Unknown   \n",
       "0  Acropora millepora   GBFSE20474-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE25693-24         Australia   \n",
       "0  Acropora millepora   GBFSE26948-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE28464-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE31107-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE35612-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE43179-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE45768-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE47451-24     Unrecoverable   \n",
       "0  Acropora millepora    GBFSE4891-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE50146-24  Papua New Guinea   \n",
       "0  Acropora millepora   GBFSE53137-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE67085-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE73842-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE74315-24  Papua New Guinea   \n",
       "0  Acropora millepora    GBFSE7881-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSE87666-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSF17245-24  Papua New Guinea   \n",
       "0  Acropora millepora   GBFSF54948-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSF54992-24     Unrecoverable   \n",
       "0  Acropora millepora   GBFSF71523-24     Unrecoverable   \n",
       "0  Acropora millepora   GBINV35301-23     Unrecoverable   \n",
       "0  Acropora millepora   GBINV40195-24         Indonesia   \n",
       "0  Acropora millepora   GBINV40196-24         Indonesia   \n",
       "0  Acropora millepora    GBMNE5136-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5137-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5138-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5139-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5140-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5141-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5142-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5143-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5144-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5145-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5146-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5147-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5148-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5149-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5150-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5151-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5152-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5153-21           Unknown   \n",
       "0  Acropora millepora    GBMNE5154-21           Unknown   \n",
       "0  Acropora millepora     OGLA1208-11         Australia   \n",
       "0  Acropora millepora     OGLA1851-11         Australia   \n",
       "0  Acropora millepora       SCL034-22          Malaysia   \n",
       "\n",
       "                                            sequence  \n",
       "0  TACGTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...  \n",
       "0  TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAATTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTTTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACACGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GTTTTTACCCGGTTATTCGGGTGAGTTGGAGTGGTCGCGGCCTGCC...  \n",
       "0  TGCTTCTACCCATGGAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTTTACCCAGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TGCTCTACCCATTCAAGTGGGTGAGTTGGAGTGGTCGCGGCCTGCG...  \n",
       "0  TGCTTTTACCCGGTTACTTGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  TGCTTCTACCCAGTTACATGGGTGAGTTGGAGTGGTCGCGGCCTGC...  \n",
       "0  GCTTTTACCCATTGAGGTGGGTGAGTTGGAGTGGTTGCGGCCTGCA...  \n",
       "0  TTGGAGTTCGGGCCGCACGCGCGCTACACTGAGCGCCGCAGCGAGT...  \n",
       "0  TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...  \n",
       "0  TAATTCT----AGAGCTAATAC--ATGCAAACA-GATT-CCGACCA...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  TTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCTCC...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  GATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGCT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  ATGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGG...  \n",
       "0  AAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAGGCAT...  \n",
       "0  TGATTGGCACGGCCTTCAGTATGTTAATAAGATTAGAGCTCTCGGC...  \n",
       "0  ---GTTATATTTAGTCTTTGGGATTGGGGCAGGCATGATTGGCACG...  \n",
       "0  ---------------------------GGCATGATTGGCACGGCCT...  \n",
       "0  CATAAAGATATTGGTACGTTATATTTAGTCTTTGGGATTGGGGCAG...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading: favia fragum\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>BOLD_process_id</th>\n",
       "      <th>country</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Favia fragum</td>\n",
       "      <td>GBCI0001-06</td>\n",
       "      <td>Panama</td>\n",
       "      <td>----------------------------------------------...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Favia fragum</td>\n",
       "      <td>GBCI0002-06</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>----------------------------------------------...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Favia fragum</td>\n",
       "      <td>GBCI0234-06</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>---------------------------------------ACTTTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Favia fragum</td>\n",
       "      <td>GBCI0235-06</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>---------------------------------------ACTTTAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        species BOLD_process_id  country  \\\n",
       "0  Favia fragum     GBCI0001-06   Panama   \n",
       "0  Favia fragum     GBCI0002-06   Brazil   \n",
       "0  Favia fragum     GBCI0234-06  Unknown   \n",
       "0  Favia fragum     GBCI0235-06  Unknown   \n",
       "\n",
       "                                            sequence  \n",
       "0  ----------------------------------------------...  \n",
       "0  ----------------------------------------------...  \n",
       "0  ---------------------------------------ACTTTAT...  \n",
       "0  ---------------------------------------ACTTTAT...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading: pocillopora damicornis\n"
     ]
    }
   ],
   "source": [
    "from bold import boldAPI\n",
    "import importlib\n",
    "importlib.reload(boldAPI)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "acropora millepora\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>BOLD_process_id</th>\n",
       "      <th>country</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [species, BOLD_process_id, country, sequence]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boldAPI.get_sequences('acropora millepora')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>BOLDProcessId</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abscondita terminalis</td>\n",
       "      <td>GBCL39854-19</td>\n",
       "      <td>CTTAATAAATGACTTTTTTCAACAAATCATAAAGACATTGGAACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abscondita terminalis</td>\n",
       "      <td>SPMIS002-22</td>\n",
       "      <td>AACTTTATATTTTATCTTCGGAATCTGAGCAGGAATATTAGGAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Actias luna luna</td>\n",
       "      <td>GBGLS176-19</td>\n",
       "      <td>TTTTGATTTTTTGGGCATCCAGAAGTTTATATTTTAATTTTACCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actias luna</td>\n",
       "      <td>LGSM242-04</td>\n",
       "      <td>AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Actias luna</td>\n",
       "      <td>LGSMC795-05</td>\n",
       "      <td>AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41944</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMTG5022-16</td>\n",
       "      <td>ATGTTAATATTTTTTTATTTTTCTACCTTTTTTCTTTGTGGTTTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41945</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMTG5022-16</td>\n",
       "      <td>ATGAATAAACCCATACGAAAAAACCACCCACTAATCAAAATTATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41946</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>GBMH10124-19</td>\n",
       "      <td>GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41947</th>\n",
       "      <td>Zootermopsis nevadensis nuttingi</td>\n",
       "      <td>GBMH9919-19</td>\n",
       "      <td>GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41948</th>\n",
       "      <td>Zootermopsis nevadensis</td>\n",
       "      <td>SPMIS419-22</td>\n",
       "      <td>AACATTATACTTCATCTTCGGGGCATGATCCGGAATATTAGGAACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40565 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Species BOLDProcessId  \\\n",
       "0                 Abscondita terminalis  GBCL39854-19   \n",
       "1                 Abscondita terminalis   SPMIS002-22   \n",
       "2                      Actias luna luna   GBGLS176-19   \n",
       "3                           Actias luna    LGSM242-04   \n",
       "4                           Actias luna   LGSMC795-05   \n",
       "...                                 ...           ...   \n",
       "41944           Zootermopsis nevadensis  GBMTG5022-16   \n",
       "41945           Zootermopsis nevadensis  GBMTG5022-16   \n",
       "41946           Zootermopsis nevadensis  GBMH10124-19   \n",
       "41947  Zootermopsis nevadensis nuttingi   GBMH9919-19   \n",
       "41948           Zootermopsis nevadensis   SPMIS419-22   \n",
       "\n",
       "                                                Sequence  \n",
       "0      CTTAATAAATGACTTTTTTCAACAAATCATAAAGACATTGGAACTT...  \n",
       "1      AACTTTATATTTTATCTTCGGAATCTGAGCAGGAATATTAGGAACC...  \n",
       "2      TTTTGATTTTTTGGGCATCCAGAAGTTTATATTTTAATTTTACCAG...  \n",
       "3      AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...  \n",
       "4      AACTTTATATTTTATTTTTGGAATTTGAGCAGGAATAGTAGGAACT...  \n",
       "...                                                  ...  \n",
       "41944  ATGTTAATATTTTTTTATTTTTCTACCTTTTTTCTTTGTGGTTTGT...  \n",
       "41945  ATGAATAAACCCATACGAAAAAACCACCCACTAATCAAAATTATAA...  \n",
       "41946  GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...  \n",
       "41947  GAACAGAACTTGGACAACCAGGATCCTAATTGGAGACGACCAAATT...  \n",
       "41948  AACATTATACTTCATCTTCGGGGCATGATCCGGAATATTAGGAACA...  \n",
       "\n",
       "[40565 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# get amino acid sequences for the DNA seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from plmcluster import proteinextraction\n",
    "# sequencePath = os.path.join('data', 'BOLD_sequences.csv')\n",
    "# if os.path.exists(sequencePath) and:\n",
    "#     results = pd.read_csv(sequencePath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jnicolow/.conda/envs/numt/lib/python3.11/site-packages/Bio/Seq.py:2334: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "results.drop_duplicates(subset='Sequence', inplace=True) # there is a chance that after N and - being removed there are sequences that are duplicates now\n",
    "results['AminoAcidSeq'] = results['Sequence'].apply(proteinextraction.get_aminoacid_seq)\n",
    "results.to_csv(os.path.join('data', 'BOLD_seqs_with_protein.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
